File: scripts/cluster/intra_connector.py
import heapq
import json
import sqlite3
from typing import Callable, Dict, List, Optional, Sequence, Set, Tuple

from .neighbor_policy import NeighborPolicy
from .jps_accelerator import expand_with_jps, extract_waypoints


def _chunk_bounds(chunk_x: int, chunk_z: int, size: int) -> Tuple[int, int, int, int]:
    x0 = chunk_x * size
    y0 = chunk_z * size
    x1 = x0 + size - 1
    y1 = y0 + size - 1
    return x0, y0, x1, y1


def _in_bounds(x: int, y: int, bounds: Tuple[int, int, int, int]) -> bool:
    x0, y0, x1, y1 = bounds
    return x0 <= x <= x1 and y0 <= y <= y1


def _is_walkable(conn: sqlite3.Connection, x: int, y: int, plane: int) -> bool:
    row = conn.execute(
        "SELECT blocked, walk_mask FROM tiles WHERE x=? AND y=? AND plane=?",
        (x, y, plane),
    ).fetchone()
    if row is None:
        return False
    blocked = int(row[0]) if row[0] is not None else 1
    walk_mask = int(row[1]) if row[1] is not None else 0
    return (blocked == 0) and (walk_mask != 0)


def _build_local_walkable(
    conn: sqlite3.Connection, chunk_x: int, chunk_z: int, plane: int
) -> Set[Tuple[int, int]]:
    rows = conn.execute(
        "SELECT x, y, blocked, walk_mask FROM tiles WHERE chunk_x=? AND chunk_z=? AND plane=?",
        (chunk_x, chunk_z, plane),
    ).fetchall()
    walkable: Set[Tuple[int, int]] = set()
    for r in rows:
        x, y = int(r[0]), int(r[1])
        blocked = int(r[2]) if r[2] is not None else 1
        walk_mask = int(r[3]) if r[3] is not None else 0
        if blocked == 0 and walk_mask != 0:
            walkable.add((x, y))
    return walkable


def _make_walkable_cb(
    conn: sqlite3.Connection,
    bounds: Tuple[int, int, int, int],
    plane: int,
    local_walkable: Set[Tuple[int, int]],
) -> Callable[[int, int, int], bool]:
    def cb(x: int, y: int, p: int) -> bool:
        if p != plane:
            # Different plane: fallback to DB
            return _is_walkable(conn, x, y, p)
        if _in_bounds(x, y, bounds):
            return (x, y) in local_walkable
        # Outside this chunk: fallback to DB
        return _is_walkable(conn, x, y, p)

    return cb


def _chebyshev_cost(ax: int, ay: int, bx: int, by: int) -> int:
    return max(abs(bx - ax), abs(by - ay))


def _a_star(
    conn: sqlite3.Connection,
    policy: NeighborPolicy,
    start: Tuple[int, int, int],
    goal: Tuple[int, int, int],
    bounds: Tuple[int, int, int, int],
    use_jps: bool,
    is_walkable_cb: Optional[Callable[[int, int, int], bool]] = None,
) -> Optional[Tuple[int, List[Tuple[int, int, int]]]]:
    sx, sy, sp = start
    gx, gy, gp = goal
    if sp != gp:
        return None
    if not (_in_bounds(sx, sy, bounds) and _in_bounds(gx, gy, bounds)):
        return None
    chk = (is_walkable_cb or (lambda x, y, p: _is_walkable(conn, x, y, p)))
    if not (chk(sx, sy, sp) and chk(gx, gy, gp)):
        return None

    open_heap: List[Tuple[int, Tuple[int, int, int]]] = []
    heapq.heappush(open_heap, (0, start))

    g_score: Dict[Tuple[int, int, int], int] = {start: 0}
    f_score: Dict[Tuple[int, int, int], int] = {start: _chebyshev_cost(sx, sy, gx, gy)}
    came: Dict[Tuple[int, int, int], Tuple[int, int, int]] = {}
    closed: set[Tuple[int, int, int]] = set()

    pol = policy if is_walkable_cb is None else policy.with_walkable(is_walkable_cb)

    while open_heap:
        _, current = heapq.heappop(open_heap)
        if current in closed:
            continue
        closed.add(current)
        cx, cy, cp = current
        if current == goal:
            # reconstruct
            path: List[Tuple[int, int, int]] = [current]
            while current in came:
                current = came[current]
                path.append(current)
            path.reverse()
            return g_score[(gx, gy, gp)], path

        # neighbors
        if use_jps:
            nbrs = expand_with_jps(conn, pol, cx, cy, cp, is_walkable_cb)
        else:
            nbrs = pol.neighbors(cx, cy, cp)

        for nx, ny, np in nbrs:
            if np != cp:
                continue
            if not _in_bounds(nx, ny, bounds):
                continue
            if (nx, ny, np) in closed:
                continue
            step = _chebyshev_cost(cx, cy, nx, ny)
            if step <= 0:
                continue
            tentative = g_score[(cx, cy, cp)] + step
            if tentative < g_score.get((nx, ny, np), 1 << 60):
                came[(nx, ny, np)] = (cx, cy, cp)
                g_score[(nx, ny, np)] = tentative
                f = tentative + _chebyshev_cost(nx, ny, gx, gy)
                f_score[(nx, ny, np)] = f
                heapq.heappush(open_heap, (f, (nx, ny, np)))

    return None


def _pairs(entrances: List[Tuple[int, int, int, int]]) -> List[Tuple[int, int]]:
    # entrances: list of (entrance_id, x, y, plane)
    n = len(entrances)
    out: List[Tuple[int, int]] = []
    for i in range(n):
        for j in range(i + 1, n):
            out.append((i, j))
    return out


def _delete_existing_scope(
    conn: sqlite3.Connection,
    chunk_x: int,
    chunk_z: int,
    plane: int,
) -> None:
    conn.execute(
        "DELETE FROM cluster_intraconnections WHERE chunk_x_from=? AND chunk_z_from=? AND plane_from=?",
        (chunk_x, chunk_z, plane),
    )


def run(
    *,
    conn: sqlite3.Connection,
    planes: Optional[Sequence[int]],
    chunk_range: Tuple[Optional[int], Optional[int], Optional[int], Optional[int]],
    recompute: bool,
    dry_run: bool,
    workers: int,
    store_paths: bool,
    logger,
) -> Dict[str, int]:
    # Load policy once per phase
    policy = NeighborPolicy.from_db(conn)

    x_min, x_max, z_min, z_max = chunk_range
    conds = ["1=1"]
    params: List[int] = []
    if x_min is not None:
        conds.append("chunk_x >= ?")
        params.append(int(x_min))
    if x_max is not None:
        conds.append("chunk_x <= ?")
        params.append(int(x_max))
    if z_min is not None:
        conds.append("chunk_z >= ?")
        params.append(int(z_min))
    if z_max is not None:
        conds.append("chunk_z <= ?")
        params.append(int(z_max))

    chunks_sql = f"SELECT chunk_x, chunk_z, chunk_size FROM chunks WHERE {' AND '.join(conds)}"
    chunks = conn.execute(chunks_sql, params).fetchall()

    total_pairs = 0
    created = 0
    solved = 0
    i = 0

    began = False
    try:
        if not dry_run:
            conn.execute("BEGIN IMMEDIATE")
            began = True

        for crow in chunks:
            i += 1
            chunk_x = int(crow[0])
            chunk_z = int(crow[1])
            size = int(crow[2])
            bounds = _chunk_bounds(chunk_x, chunk_z, size)

            print(f"Processing chunk {chunk_x},{chunk_z} ({size}x{size}) {i}/ {len(chunks)} total chunks")

            # planes in this chunk scope
            if planes is None:
                p_rows = conn.execute(
                    "SELECT DISTINCT plane FROM tiles WHERE chunk_x=? AND chunk_z=?",
                    (chunk_x, chunk_z),
                ).fetchall()
                planes_local = [int(r[0]) for r in p_rows]
            else:
                planes_local = list(planes)

            for plane in planes_local:
                # entrances in this chunk+plane
                erows = conn.execute(
                    """
                    SELECT entrance_id, x, y FROM cluster_entrances
                    WHERE chunk_x=? AND chunk_z=? AND plane=?
                    ORDER BY entrance_id ASC
                    """,
                    (chunk_x, chunk_z, plane),
                ).fetchall()
                entrances = [(int(r[0]), int(r[1]), int(r[2]), plane) for r in erows]
                if not entrances:
                    continue

                # Build local walkability cache and callback
                local_walkable = _build_local_walkable(conn, chunk_x, chunk_z, plane)
                walk_cb = _make_walkable_cb(conn, bounds, plane, local_walkable)

                if recompute and not dry_run:
                    _delete_existing_scope(conn, chunk_x, chunk_z, plane)

                pairs = _pairs(entrances)
                total_pairs += len(pairs)
                for i, j in pairs:
                    e1 = entrances[i]
                    e2 = entrances[j]
                    eid1, x1, y1, p1 = e1
                    eid2, x2, y2, p2 = e2

                    # Skip if same tile
                    if x1 == x2 and y1 == y2 and p1 == p2:
                        continue

                    # Plan path (prefer JPS expansion when available)
                    use_jps = True
                    found = _a_star(
                        conn,
                        policy,
                        (x1, y1, p1),
                        (x2, y2, p2),
                        bounds,
                        use_jps,
                        walk_cb,
                    )
                    if found is None:
                        # Non-reachable pair
                        continue
                    cost, path = found
                    solved += 1

                    waypoint_blob = None
                    if store_paths:
                        waypoints = extract_waypoints(path)
                        # encode waypoints as compact JSON [[x,y],[x,y],...]
                        packed = [[px, py] for (px, py, _) in waypoints]
                        waypoint_blob = json.dumps(packed).encode("utf-8")

                    if dry_run:
                        created += 2  # both directions recorded
                        continue

                    # Insert both directions, idempotent
                    conn.execute(
                        """
                        INSERT INTO cluster_intraconnections
                          (chunk_x_from, chunk_z_from, plane_from, entrance_from, entrance_to, cost, path_blob)
                        VALUES (?, ?, ?, ?, ?, ?, ?)
                        ON CONFLICT(chunk_x_from, chunk_z_from, plane_from, entrance_from, entrance_to)
                        DO UPDATE SET
                          cost = MIN(cluster_intraconnections.cost, excluded.cost),
                          path_blob = CASE WHEN excluded.path_blob IS NOT NULL THEN excluded.path_blob ELSE cluster_intraconnections.path_blob END
                        """,
                        (chunk_x, chunk_z, plane, eid1, eid2, cost, waypoint_blob),
                    )
                    conn.execute(
                        """
                        INSERT INTO cluster_intraconnections
                          (chunk_x_from, chunk_z_from, plane_from, entrance_from, entrance_to, cost, path_blob)
                        VALUES (?, ?, ?, ?, ?, ?, ?)
                        ON CONFLICT(chunk_x_from, chunk_z_from, plane_from, entrance_from, entrance_to)
                        DO UPDATE SET
                          cost = MIN(cluster_intraconnections.cost, excluded.cost),
                          path_blob = CASE WHEN excluded.path_blob IS NOT NULL THEN excluded.path_blob ELSE cluster_intraconnections.path_blob END
                        """,
                        (chunk_x, chunk_z, plane, eid2, eid1, cost, waypoint_blob),
                    )
                    created += 2

        if began:
            conn.execute("COMMIT")
    except Exception:
        if began:
            conn.execute("ROLLBACK")
        raise

    logger.info(
        "intra_connector: pairs_total=%s pairs_solved=%s rows_created=%s dry_run=%s",
        total_pairs,
        solved,
        created,
        dry_run,
    )

    return {"created": created, "pairs_total": total_pairs, "pairs_solved": solved}

File: rust/src/commands/cluster/intra_connector.rs
use anyhow::Result;
use rusqlite::{params, Connection, OptionalExtension};
use std::cmp::Ordering;
use std::collections::{BTreeSet, HashMap, HashSet};

use super::config::Config;
use super::db::{with_tx};
use super::neighbor_policy::{MovementPolicy, Offset};

#[derive(Clone, Debug, Default)]
pub struct IntraStats {
    pub clusters_processed: usize,
    pub edges_created: usize,
}

pub fn build_intra_edges(tiles_db: &Connection, out_db: &mut Connection, cfg: &Config) -> Result<IntraStats> {
    // ensure_schema(out_db)?;

    let mut stats = IntraStats::default();

    // Collect clusters that have at least 2 entrances
    let mut clusters: Vec<(i64,i32)> = Vec::new();
    {
        let mut q = out_db.prepare(
            "SELECT c.cluster_id, c.plane, COUNT(ce.entrance_id) AS ecnt
             FROM clusters c
             JOIN cluster_entrances ce ON ce.cluster_id = c.cluster_id
             GROUP BY c.cluster_id, c.plane
             HAVING ecnt >= 2"
        )?;
        let rows = q.query_map([], |row| {
            let cluster_id: i64 = row.get(0)?;
            let plane: i32 = row.get(1)?;
            Ok((cluster_id, plane))
        })?;
        for r in rows { clusters.push(r?); }
    }

    // Filter by cfg (by plane only here; range filtering is applied per-entrance later)
    clusters.retain(|&(_cid, plane)| {
        if let Some(planes) = &cfg.planes { if !planes.contains(&plane) { return false; } }
        true
    });

    // Deterministic order
    clusters.sort_unstable();

    println!("intra: found {} clusters with >=2 entrances", clusters.len());

    // Movement and costs
    let policy = MovementPolicy::default();
    let offsets = policy.neighbor_offsets();
    let (cost_card, cost_diag) = load_movement_costs(out_db)?;
    println!(
        "intra: movement costs loaded -> straight={} diagonal={}",
        cost_card, cost_diag
    );

    // Track clusters we have cleared to avoid repeated deletes
    let mut cleared_clusters: BTreeSet<i64> = BTreeSet::new();

    if cfg.dry_run {
        println!("intra: dry_run enabled - skipping database writes");
    } else {
        println!("intra: intra edge generation will commit after each cluster");
    }

    for (cluster_id, plane) in clusters.iter().copied() {
        println!("intra: processing cluster {} on plane {}", cluster_id, plane);
        if cfg.dry_run {
            continue;
        }

        let processed_edges = with_tx(out_db, |tx| {
            if !cleared_clusters.contains(&cluster_id) {
                println!("intra: clearing existing intraconnections for cluster {}", cluster_id);
                tx.execute(
                    "DELETE FROM cluster_intraconnections \
                     WHERE entrance_from IN (SELECT entrance_id FROM cluster_entrances WHERE cluster_id=?1)",
                    params![cluster_id],
                )?;
                cleared_clusters.insert(cluster_id);
            }

            // Get entrances for this cluster (filter by chunk range if provided)
            let mut es = tx.prepare(
                "SELECT entrance_id, x, y, neighbor_dir FROM cluster_entrances WHERE cluster_id=?1 ORDER BY entrance_id"
            )?;
            let erows = es.query_map(params![cluster_id], |row| {
                let eid: i64 = row.get(0)?;
                let x: i32 = row.get(1)?;
                let y: i32 = row.get(2)?;
                let d: String = row.get(3)?;
                Ok((eid, x, y, d.chars().next().unwrap_or('?')))
            })?;
            let mut entrances: Vec<(i64,i32,i32,char)> = Vec::new();
            for r in erows { entrances.push(r?); }
            println!(
                "intra: cluster {} has {} entrances",
                cluster_id,
                entrances.len()
            );
            if entrances.len() < 2 {
                println!("intra: skipping cluster {} (needs >=2 entrances)", cluster_id);
                return Ok(None);
            }

            // Optional chunk-range filter: keep cluster only if at least one entrance lies within range
            if let Some((xmin,xmax,zmin,zmax)) = cfg.chunk_range {
                if !entrances.iter().any(|&(_eid, x, y, _)| {
                    let cx = x >> 6; let cz = y >> 6; cx >= xmin && cx <= xmax && cz >= zmin && cz <= zmax
                }) {
                    println!(
                        "intra: skipping cluster {} (no entrances within configured chunk range)",
                        cluster_id
                    );
                    return Ok(None);
                }
            }

            // Load cluster tile set
            let tile_set = load_cluster_tiles(&tx, cluster_id, plane)?;
            println!(
                "intra: cluster {} tile set size = {}",
                cluster_id,
                tile_set.len()
            );

            // Quick membership check for entrances
            if !entrances.iter().all(|&(_eid, x, y, _)| tile_set.contains(&(x,y))) {
                println!("intra: skipping cluster {} (entrance outside cluster tiles)", cluster_id);
                return Ok(None);
            }

            // Pre-prepare insert statement
            let mut ins = tx.prepare(
                "INSERT INTO cluster_intraconnections (entrance_from, entrance_to, cost, path_blob)
                 VALUES (?1,?2,?3,?4)
                 ON CONFLICT(entrance_from, entrance_to)
                 DO UPDATE SET cost = MIN(cluster_intraconnections.cost, excluded.cost),
                               path_blob = COALESCE(cluster_intraconnections.path_blob, excluded.path_blob)"
            )?;

            // Precompute external exit cluster per entrance to allow skipping redundant pairs.
            // External means the neighbor tile belongs to a different cluster than current.
            let mut q_exit = tx.prepare(
                "SELECT cluster_id FROM cluster_tiles WHERE x=?1 AND y=?2 AND plane=?3 LIMIT 1"
            )?;
            let mut exit_clusters: Vec<Option<i64>> = Vec::with_capacity(entrances.len());
            for &(_eid, x, y, d) in entrances.iter() {
                let (dx, dy) = match d {
                    'N' => (0, 1),
                    'S' => (0, -1),
                    'E' => (1, 0),
                    'W' => (-1, 0),
                    _ => (0, 0),
                };
                if dx == 0 && dy == 0 {
                    exit_clusters.push(None);
                    continue;
                }
                let nx = x + dx; let ny = y + dy;
                let cid_opt: Option<i64> = q_exit
                    .query_row(params![nx, ny, plane], |r| r.get(0))
                    .optional()?;
                let cid_opt = cid_opt.filter(|&cid| cid != cluster_id);
                exit_clusters.push(cid_opt);
            }

            // Compute all-pairs shortest paths between entrances deterministically
            let mut cluster_edge_count = 0usize;
            for i in 0..entrances.len() {
                for j in 0..entrances.len() {
                    if i == j { continue; }
                    // Skip redundant pairs where both entrances exit to the same external cluster.
                    if let (Some(ci), Some(cj)) = (exit_clusters[i], exit_clusters[j]) {
                        if ci == cj {
                            let (eid_a, _, _, _ ) = entrances[i];
                            let (eid_b, _, _, _ ) = entrances[j];
                            println!(
                                "intra: skipping pair within cluster {} because entrances {} and {} both exit to cluster {}",
                                cluster_id, eid_a, eid_b, ci
                            );
                            continue;
                        }
                    }
                    let (eid_a, ax, ay, _da) = entrances[i];
                    let (eid_b, bx, by, _db) = entrances[j];
                    // Deterministic ordering via indices already
                    println!(
                        "intra: computing path cluster {} entrance {} -> {}",
                        cluster_id,
                        eid_a,
                        eid_b
                    );
                    let (cost, path_opt) = shortest_path_in_set((ax,ay), (bx,by), &tile_set, offsets, cost_card, cost_diag);
                    if let Some(total) = cost {
                        let blob = if cfg.store_paths { path_opt.map(|p| encode_path_blob(p, plane)) } else { None };
                        ins.execute(params![eid_a, eid_b, total as i64, blob])?;
                        println!(
                            "intra: stored intra edge {} -> {} cost={} (blob={})",
                            eid_a,
                            eid_b,
                            total,
                            blob.as_ref().map(|b| b.len()).unwrap_or(0)
                        );
                        cluster_edge_count += 1;
                    } else {
                        println!(
                            "intra: no path found within cluster {} from entrance {} to {}",
                            cluster_id,
                            eid_a,
                            eid_b
                        );
                    }
                }
            }

            println!(
                "intra: completed cluster {} -> {} intra edges inserted",
                cluster_id,
                cluster_edge_count
            );
            Ok(Some(cluster_edge_count))
        })?;

        if let Some(edge_count) = processed_edges {
            stats.clusters_processed += 1;
            stats.edges_created += edge_count;
        }
    }

    println!(
        "intra: build complete (dry_run={}) -> clusters_processed={} edges_created={}",
        cfg.dry_run,
        stats.clusters_processed,
        stats.edges_created
    );
    Ok(stats)
}

fn load_cluster_tiles(tx: &rusqlite::Transaction<'_>, cluster_id: i64, plane: i32) -> Result<HashSet<(i32,i32)>> {
    let mut tiles_stmt = tx.prepare(
        "SELECT x, y FROM cluster_tiles WHERE cluster_id=?1 AND plane=?2",
    )?;
    let rows = tiles_stmt.query_map(params![cluster_id, plane], |row| {
        let x: i32 = row.get(0)?;
        let y: i32 = row.get(1)?;
        Ok((x,y))
    })?;
    let mut set: HashSet<(i32,i32)> = HashSet::new();
    for r in rows { set.insert(r?); }
    Ok(set)
}

fn load_movement_costs(db: &Connection) -> Result<(i64,i64)> {
    use rusqlite::OptionalExtension;
    let straight: Option<String> = db
        .query_row("SELECT value FROM meta WHERE key='movement_cost_straight'", [], |r| r.get(0))
        .optional()?;
    let diagonal: Option<String> = db
        .query_row("SELECT value FROM meta WHERE key='movement_cost_diagonal'", [], |r| r.get(0))
        .optional()?;
    let sc = straight.and_then(|s| s.parse::<i64>().ok()).unwrap_or(600);
    let dc = diagonal.and_then(|s| s.parse::<i64>().ok()).unwrap_or(1000);
    Ok((sc, dc))
}

fn shortest_path_in_set(
    start: (i32,i32),
    goal: (i32,i32),
    tiles: &HashSet<(i32,i32)>,
    offsets: &[Offset],
    cost_card: i64,
    cost_diag: i64,
) -> (Option<i64>, Option<Vec<(i32,i32)>>) {
    if start == goal { return (Some(0), Some(vec![start])); }
    if !tiles.contains(&start) || !tiles.contains(&goal) { return (None, None); }

    // Dijkstra with deterministic tie-breaking using BTreeMap frontier keyed by (cost, x, y)
    let mut frontier: BTreeSet<(i64, i32, i32)> = BTreeSet::new();
    let mut dist: HashMap<(i32,i32), i64> = HashMap::new();
    let mut prev: HashMap<(i32,i32), (i32,i32)> = HashMap::new();

    dist.insert(start, 0);
    frontier.insert((0, start.0, start.1));

    while let Some((d, x, y)) = frontier.iter().next().cloned() {
        frontier.remove(&(d, x, y));
        if (x,y) == goal { break; }
        for &Offset(dx,dy) in offsets.iter() {
            let nx = x + dx; let ny = y + dy;
            let n = (nx,ny);
            if !tiles.contains(&n) { continue; }
            let step = if dx == 0 || dy == 0 { cost_card } else { cost_diag };
            let nd = d + step;
            match dist.get(&n) {
                Some(&old) if old < nd => {},
                Some(&old) if old == nd => {
                    // tie-breaker: prefer lexicographically smaller (nx,ny)
                    let better = (nx,ny).cmp(&(prev.get(&n).cloned().unwrap_or((i32::MAX,i32::MAX))));
                    if better == Ordering::Less {
                        prev.insert(n, (x,y));
                    }
                },
                _ => {
                    // update
                    if let Some(&old) = dist.get(&n) {
                        frontier.remove(&(old, nx, ny));
                    }
                    dist.insert(n, nd);
                    prev.insert(n, (x,y));
                    frontier.insert((nd, nx, ny));
                }
            }
        }
    }

    if let Some(&total) = dist.get(&goal) {
        // reconstruct path
        let mut path: Vec<(i32,i32)> = Vec::new();
        let mut cur = goal;
        path.push(cur);
        while cur != start {
            if let Some(&p) = prev.get(&cur) { cur = p; path.push(cur); } else { break; }
        }
        path.reverse();
        (Some(total), Some(path))
    } else {
        (None, None)
    }
}

fn encode_path_blob(path: Vec<(i32,i32)>, plane: i32) -> Vec<u8> {
    let reduced = reduce_path_to_breakpoints(&path);
    let mut out = Vec::with_capacity(reduced.len() * 12);
    let plane_bytes = plane.to_le_bytes();
    for (x,y) in reduced {
        out.extend_from_slice(&x.to_le_bytes());
        out.extend_from_slice(&y.to_le_bytes());
        out.extend_from_slice(&plane_bytes);
    }
    out
}

fn reduce_path_to_breakpoints(path: &[(i32,i32)]) -> Vec<(i32,i32)> {
    match path.len() {
        0 => Vec::new(),
        1 => path.to_vec(),
        _ => {
            let mut reduced: Vec<(i32,i32)> = Vec::with_capacity(path.len());
            reduced.push(path[0]);
            for window in path.windows(3) {
                let prev = window[0];
                let cur = window[1];
                let next = window[2];
                let dir_in = movement_dir(prev, cur);
                let dir_out = movement_dir(cur, next);
                if dir_in != dir_out {
                    if reduced.last().copied() != Some(cur) {
                        reduced.push(cur);
                    }
                }
            }
            if let Some(&last) = path.last() {
                if reduced.last().copied() != Some(last) {
                    reduced.push(last);
                }
            }
            reduced
        }
    }
}

fn movement_dir(from: (i32,i32), to: (i32,i32)) -> Option<(i32,i32)> {
    let dx = to.0 - from.0;
    let dy = to.1 - from.1;
    if dx == 0 && dy == 0 {
        None
    } else {
        Some((dx.signum(), dy.signum()))
    }
}

// deterministic_cluster_id no longer needed with explicit cluster IDs in DB

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::NamedTempFile;

    fn setup_small_cluster() -> Result<(Connection, Connection, Config, (i64, (i32,i32), (i32,i32)))> {
        let tiles_dbf = NamedTempFile::new().unwrap();
        let out_dbf = NamedTempFile::new().unwrap();
        let mut tiles = Connection::open(tiles_dbf.path())?;
        let mut out = Connection::open(out_dbf.path())?;
        crate::db::create_tables(&mut tiles)?;
        crate::db::create_tables(&mut out)?;

        // A 2-tile line, plane=0
        tiles.execute("INSERT INTO tiles(x,y,plane,flag,blocked,walk_mask,blocked_mask,walk_data) VALUES (0,0,0,0,0,1,0,'')", [])?;
        tiles.execute("INSERT INTO tiles(x,y,plane,flag,blocked,walk_mask,blocked_mask,walk_data) VALUES (1,0,0,0,0,1,0,'')", [])?;

        // Create a cluster with two tiles and two entrances
        let cid: i64 = 1;
        out.execute("INSERT INTO clusters(cluster_id, plane, label, tile_count) VALUES (?1,0,0,2)", params![cid])?;
        out.execute("INSERT INTO cluster_tiles(cluster_id, x, y, plane) VALUES (?1,0,0,0)", params![cid])?;
        out.execute("INSERT INTO cluster_tiles(cluster_id, x, y, plane) VALUES (?1,1,0,0)", params![cid])?;
        out.execute("INSERT INTO cluster_entrances(cluster_id, x, y, plane, neighbor_dir) VALUES (?1,0,0,0,'E')", params![cid])?;
        out.execute("INSERT INTO cluster_entrances(cluster_id, x, y, plane, neighbor_dir) VALUES (?1,1,0,0,'W')", params![cid])?;

        let cfg = Config::default();
        Ok((tiles, out, cfg, (cid, (0,0), (1,0))))
    }

    #[test]
    fn intra_edges_no_blobs_when_flag_false() -> Result<()> {
        let (tiles, mut out, mut cfg, _info) = setup_small_cluster()?;
        cfg.store_paths = false;
        let stats = build_intra_edges(&tiles, &mut out, &cfg)?;
        assert_eq!(stats.clusters_processed, 1);
        // Expect two directed edges (A->B and B->A)
        let cnt: i64 = out.query_row("SELECT COUNT(*) FROM cluster_intraconnections", [], |r| r.get(0))?;
        assert_eq!(cnt, 2);
        let n_blobs: i64 = out.query_row("SELECT COUNT(*) FROM cluster_intraconnections WHERE path_blob IS NOT NULL", [], |r| r.get(0))?;
        assert_eq!(n_blobs, 0);
        Ok(())
    }

    #[test]
    fn intra_edges_with_blobs_when_flag_true() -> Result<()> {
        let (tiles, mut out, mut cfg, _info) = setup_small_cluster()?;
        cfg.store_paths = true;
        let stats = build_intra_edges(&tiles, &mut out, &cfg)?;
        assert_eq!(stats.clusters_processed, 1);
        let cnt: i64 = out.query_row("SELECT COUNT(*) FROM cluster_intraconnections", [], |r| r.get(0))?;
        assert_eq!(cnt, 2);
        let n_blobs: i64 = out.query_row("SELECT COUNT(*) FROM cluster_intraconnections WHERE path_blob IS NOT NULL", [], |r| r.get(0))?;
        assert_eq!(n_blobs, 2);
        Ok(())
    }

    #[test]
    fn encode_path_blob_reduces_straight_segments() {
        let path = vec![(0, 0), (0, 1), (0, 2), (1, 2), (2, 2)];
        let blob = encode_path_blob(path.clone(), 0);
        assert_eq!(blob.len(), 3 * 12);

        let decoded: Vec<(i32, i32)> = blob
            .chunks_exact(12)
            .map(|chunk| {
                let (xb, yb) = chunk.split_at(4);
                let (yb, _) = yb.split_at(4);
                (
                    i32::from_le_bytes(xb.try_into().unwrap()),
                    i32::from_le_bytes(yb.try_into().unwrap()),
                )
            })
            .collect();
        assert_eq!(decoded, vec![(0, 0), (0, 2), (2, 2)]);
    }

    #[test]
    fn encode_path_blob_preserves_diagonal_runs() {
        let path = vec![(0, 0), (1, 1), (2, 2), (3, 3)];
        let blob = encode_path_blob(path.clone(), 0);
        assert_eq!(blob.len(), 2 * 12);

        let decoded: Vec<(i32, i32)> = blob
            .chunks_exact(12)
            .map(|chunk| {
                let (xb, yb) = chunk.split_at(4);
                let (yb, _) = yb.split_at(4);
                (
                    i32::from_le_bytes(xb.try_into().unwrap()),
                    i32::from_le_bytes(yb.try_into().unwrap()),
                )
            })
            .collect();
        assert_eq!(decoded, vec![(0, 0), (3, 3)]);
    }
}
